feat(replays): Add ReplayVideo envelope item type

cmanallen opened the pull request:
Add support for replay-video event types.

Changes:

- Refactors event validation in the processor. This is to DRY the processing logic between the three replay event types.
- Adds new replay_video field to replay-recording Kafka message.
- Add outcomes.
- Configure rate-limits.

Relates to:
 * https://github.com/getsentry/sentry/issues/63255

JoshFerge committed:
feat(replay): combined envelope items

JoshFerge committed:
only run in processing mode

JoshFerge committed:
add integration test. fix processing mode stuff, add store logic

JoshFerge committed:
remove processing mode business, use vec instead of Bytes

JoshFerge committed:
fix args in combined store func

JoshFerge committed:
fix tests

JoshFerge committed:
remove unused test

JoshFerge committed:
merge

JoshFerge committed:
get it compiling

JoshFerge committed:
get tests running

JoshFerge committed:
remove unused imports

JoshFerge committed:
make clippy even more happy

JoshFerge committed:
add changelog entry

JoshFerge committed:
move logic to store

JoshFerge committed:
fix

JoshFerge committed:
fix merge

JoshFerge committed:
more cleanup

JoshFerge committed:
Update CHANGELOG.md

Co-authored-by: Joris Bayer <joris.bayer@sentry.io>

JoshFerge committed:
remove feature

JoshFerge committed:
use flag in header

JoshFerge committed:
merge

JoshFerge committed:
try to debug

JoshFerge committed:
tests working

JoshFerge committed:
dont produce two messages

JoshFerge committed:
cleanup

JoshFerge committed:
remove println

JoshFerge committed:
fix lint

JoshFerge committed:
combined_payload not used in non-processing, allow dead code

JoshFerge committed:
Update relay-server/src/envelope.rs

Co-authored-by: Joris Bayer <joris.bayer@sentry.io>

JoshFerge committed:
simplify processing feature declare on env item flag

JoshFerge committed:
move logic into one func

JoshFerge committed:
clean refactor

JoshFerge committed:
add one final test / fix impl

JoshFerge committed:
small refactor to produce function

JoshFerge committed:
Merge branch 'master' into jferg/combine-replay-env-items

cmanallen committed:
Merge branch 'master' into cmanallen/replays-add-video-envelope-item

cmanallen committed:
Add ReplayVideo envelope item type

JoshFerge committed:
Merge branch 'master' into jferg/combine-replay-env-items

cmanallen committed:
Add video envelope item validator

cmanallen committed:
Add ReplayVideo item type processor

cmanallen committed:
Send optional replay_video field in Kafka payload

cmanallen committed:
Register outcome, rate-limit, and processing error

cmanallen committed:
Merge branch 'jferg/combine-replay-env-items' into cmanallen/replays-add-video-envelope-item

cmanallen committed:
Compute payload through addition

cmanallen committed:
Add docs

cmanallen committed:
Update changelog

cmanallen committed:
Re-phrase changelog entry

JoshFerge committed:
only produce replay events if replay recording / event exists

cmanallen committed:
Merge branch 'jferg/combine-replay-env-items' into cmanallen/replays-add-video-envelope-item

cmanallen committed:
Remove redundant boolean condition

cmanallen committed:
Flatten extraction and consider replay_event size when measuring payload size

cmanallen committed:
Replay-event variable is only populated if the flag is enabled

cmanallen committed:
Remove redundant method

cmanallen committed:
Remove isize note

cmanallen committed:
Correct boolean condition

cmanallen committed:
Emit outcome if the payload is too large

cmanallen committed:
Merge branch 'jferg/combine-replay-env-items' into cmanallen/replays-add-video-envelope-item

cmanallen committed:
Rename error and update comments

cmanallen committed:
Merge branch 'master' into cmanallen/replays-add-video-envelope-item

cmanallen committed:
Add replay-video test coverage

cmanallen committed:
Remove unused test case

jjbayer reviewed (commented):
> Is there a good pattern for handling envelope-items that depend on the presence of another, valid envelope-item? I.e. erroring if we see one but not the other.

There is no good pattern for that, but we do it, e.g. here: https://github.com/getsentry/relay/blob/a9baace4ae06626e29d3b1cdcb07b4113523a5d4/relay-server/src/services/processor/profile.rs#L29-L30

> How are rate-limits handled? These are logically one request but rate-limits are calculated per item.

If a video always has a recording, but a recording does not necessarily have a video, I would always drop the video with the recording (see code below), but only emit an outcome for the recording. https://github.com/getsentry/relay/blob/ac1ff144600f64f1ceac79b7298cd97b6545812b/relay-server/src/utils/rate_limits.rs#L640-L645

In general, if recordings and videos are tightly coupled, I would personally embed the video into the recording item, unless that is a problem size-wise.

cmanallen committed:
Add support for SDK merged replay-video envelope item

cmanallen committed:
Set processing items on the envelope item's headers

cmanallen committed:
Update store to accept combined ReplayVideo item type

cmanallen committed:
Fix lint

cmanallen committed:
Add ReplayVideo to ReplayGroup

cmanallen committed:
Add serde_bytes

cmanallen committed:
Add test_replay_videos test module

cmanallen committed:
Remove unused import

TBS1996 commented on relay-server/src/envelope.rs:
```diff
@@ -771,6 +783,17 @@ impl Item {
+    /// Returns the payload's replay video events.
+    #[cfg(feature = "processing")]
+    pub fn replay_video_events(&self) -> Option<(Vec<u8>, Vec<u8>)> {
+        self.headers.replay_video_events.clone()
```
clone here seems a bit too implicit, I get that we need an owned value to make a Bytes object, but maybe defer cloning to the caller so they know it's potentially expensive.  
 
we could also look into something like take_video_events if we only need to use it once and the payload is large

jjbayer commented on relay-server/src/envelope.rs:
```diff
@@ -509,6 +513,11 @@ pub struct ItemHeaders {
+    /// The parsed replay-event and replay-recording payloads.
+    /// NOTE: This is internal-only and not exposed into the Envelope.
+    #[serde(default, skip)]
+    replay_video_events: Option<(Vec<u8>, Vec<u8>)>,
```
I would prefer if we don't put this on the header, but make it part of the item payload somehow. We should either

1. manage to encode all information into a single item payload (e.g. `{"some": "json"}\x00vide0_bytes`),
2. let them be separate items after all, and only combine them in `store.rs`.

@cmanallen let me know if you would prefer to discuss this in a meeting.

cmanallen replied:
Yes I would like to discuss in a meeting.

jjbayer commented on relay-server/src/services/processor/replay.rs:
```diff
@@ -54,82 +54,108 @@ pub fn process(
 }
 
+enum ProcessingAction<T> {
+    Drop(ItemAction),
```
```suggestion
    Drop(Outcome),
```

jjbayer commented on relay-server/src/envelope.rs:
```diff
@@ -213,6 +216,7 @@ impl std::str::FromStr for ItemType {
             "profile" => Self::Profile,
             "replay_event" => Self::ReplayEvent,
             "replay_recording" => Self::ReplayRecording,
+            "replay_video" => Self::ReplayVideo,
```
How large would we expect these videos to be?

cmanallen replied:
10MB max.

cmanallen committed:
Pass drop as an Outcome and allow the outerscope to manage how that outcome is handled

cmanallen committed:
Merge branch 'master' into cmanallen/replays-add-video-envelope-item

cmanallen committed:
Revert "Pass drop as an Outcome and allow the outerscope to manage how that outcome is handled"

This reverts commit bed786b962f06c261a0ff2608d6086ec9ee054ba.

jjbayer committed:
ref(replays): video item (#3149)

Serialize processed replay video item back into the envelope instead of
putting it on the header.

#skip-changelog

cmanallen committed:
Merge branch 'cmanallen/replays-add-video-envelope-item' of https://github.com/getsentry/relay into cmanallen/replays-add-video-envelope-item

jjbayer commented on CHANGELOG.md:
```diff
@@ -28,6 +28,7 @@
 **Features**:
 
 - Add protobuf support for ingesting OpenTelemetry spans and use official `opentelemetry-proto` generated structs. ([#3044](https://github.com/getsentry/relay/pull/3044))
+- Adds ReplayVideo envelope-item type. ([#3105](https://github.com/getsentry/relay/pull/3105))
```
Please move this to the `## Unreleased` section.

jjbayer commented on relay-server/Cargo.toml:
```diff
@@ -97,6 +96,7 @@ reqwest = { version = "0.11.1", features = [
 rmp-serde = "1.1.1"
 rust-embed = { version = "8.0.0", optional = true }
 serde = { workspace = true }
+serde_bytes = "0.11"
```
This is a crate by the creators of serde, so adding it as dependency should be fine.

cmanallen committed:
Update changelog

cmanallen committed:
Merge branch 'master' into cmanallen/replays-add-video-envelope-item

cmanallen committed:
Merge branch 'master' into cmanallen/replays-add-video-envelope-item

cmanallen committed:
Produce replay-event to the snuba consumer

cmanallen committed:
Merge branch 'cmanallen/replays-add-video-envelope-item' of https://github.com/getsentry/relay into cmanallen/replays-add-video-envelope-item

cmanallen committed:
Create a new Bytes struct with a copy of the replay-event bytes

cmanallen committed:
Use copy_from_slice

cmanallen committed:
Add replay-video produced metric

cmanallen committed:
Merge branch 'master' into cmanallen/replays-add-video-envelope-item

cmanallen committed:
Abstract replay-event assertion and use in replay-video coverage

cmanallen committed:
Assert retention days

jjbayer commented on relay-server/src/services/store.rs:
```diff
@@ -891,6 +908,66 @@ impl StoreService {
+        metric!(
+            counter(RelayCounters::ProcessingMessageProduced) += 1,
+            event_type = "replay_video"
+        );
```
`produce_replay_recording` already logs a metric. We should emit 1 metric for every Kafka message, this would emit two IIURC. What we could do is change the "event type" in `produce_replay_recording` to `recording_with_video` or similar if a video is passed with it.

jjbayer commented on relay-server/src/services/store.rs:
```diff
@@ -891,6 +908,66 @@ impl StoreService {
+            start_time,
+            retention,
+            Bytes::copy_from_slice(replay_event),
+        )?;
```
Why do we need to produce the event a second time? It's already published as part of the replay recording, right?

cmanallen replied:
Its being published to a different consumer.  We're producing one event to the replay_recording consumer and another event to the replay_event consumer.  The duplicated data being sent to the recording consumer for reasons other than storing the replay_event bytes.  Refer to this PR https://github.com/getsentry/relay/pull/3035 where we merged the payloads.

In the future we will drop the `produce_replay_event` function from Relay but we're not ready yet.

jjbayer commented on relay-server/src/services/store.rs:
```diff
@@ -891,6 +908,66 @@ impl StoreService {
+            scoping.project_id,
+            start_time,
+            retention,
+            Bytes::copy_from_slice(replay_event),
```
If we really need to duplicate the event, let's change the signature of `produce_replay_event` to take a reference, so we don't copy here.

cmanallen committed:
Emit a single metric

cmanallen committed:
Remove copy
